### PR TITLE
support Active directory change user passsword

### PR DESCRIPTION
**Pull request by mickeykcyolo**

### CODE REVIEW

File: ldap-client.go [hunk 1]
```diff
@@ -244,3 +247,89 @@ func (lc *LDAPClient) GetAllGroupsByName(groupName string) ([]LdapGroup, error)
...
+	// bind as the user to verify their current password
+	if err = lc.Conn.Bind(userDN, oldPassword); err != nil {
+		err = fmt.Errorf("could not bind user via old password: %s", err.Error())
+		return
+	}
+
+	utf16 := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM)
+
+	var encodePasswordForAD = func(pass string) (encoded string, err error) {
+		if encoded, err = utf16.NewEncoder().String(fmt.Sprintf("\"%s\"", pass)); err != nil {
```
Comment (avivklas) [inline ldap-client.go:305]:
```suggestion
		if encoded, err = utf16.NewEncoder().String(fmt.Sprintf("%q", pass)); err != nil {
```

File: ldap-client.go [hunk 2]
```diff
@@ -244,3 +247,89 @@ func (lc *LDAPClient) GetAllGroupsByName(groupName string) ([]LdapGroup, error)
...
+			err = fmt.Errorf("password could not be utf16 encoded: %s", err.Error())
```
Comment (avivklas) [inline ldap-client.go:306]: wrapping is better than simply flattening the error 
```suggestion
			err = fmt.Errorf("password could not be utf16 encoded: %w", err)
```

File: ldap-client.go [hunk 3]
```diff
@@ -244,3 +247,89 @@ func (lc *LDAPClient) GetAllGroupsByName(groupName string) ([]LdapGroup, error)
...
+
+	// bind as the user to verify their current password
+	if err = lc.Conn.Bind(userDN, oldPassword); err != nil {
+		err = fmt.Errorf("could not bind user via old password: %s", err.Error())
+		return
+	}
+
+	utf16 := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM)
+
+	var encodePasswordForAD = func(pass string) (encoded string, err error) {
```
Comment (avivklas) [inline ldap-client.go:304]: why do you define this method here instead of on the struct?

File: ldap-client.go [hunk 4]
```diff
@@ -244,3 +247,89 @@ func (lc *LDAPClient) GetAllGroupsByName(groupName string) ([]LdapGroup, error)
...
+	}
+	if newEncodedPass, err = encodePasswordForAD(newPassword); err != nil {
+		return
+	}
+
+	modify := ldap.NewModifyRequest(userDN)
+	modify.Delete(adPasswordAttributeName, []string{oldEncodedPass})
+	modify.Add(adPasswordAttributeName, []string{newEncodedPass})
+	if err = lc.Conn.Modify(modify); err != nil {
+		err = fmt.Errorf("password could not be changed: %s", err.Error())
```
Comment (avivklas) [inline ldap-client.go:324]:
```suggestion
		err = fmt.Errorf("password could not be changed: %w", err)
```

File: ldap-client.go [hunk 5]
```diff
@@ -244,3 +247,89 @@ func (lc *LDAPClient) GetAllGroupsByName(groupName string) ([]LdapGroup, error)
...
+		return
+	}
+
+	// bind as the user to verify their new password
+	if err = lc.Conn.Bind(userDN, newPassword); err != nil {
+		err = fmt.Errorf("could not bind user via new password: %s", err.Error())
```
Comment (avivklas) [inline ldap-client.go:330]:
```suggestion
		err = fmt.Errorf("could not bind user via new password: %w", err)
```

**Review by avivklas — CHANGES_REQUESTED** [review]

File: ldap-client.go (hunk 3 above)
Comment (mickeykcyolo) [reply ldap-client.go:304]: why not?

**Review by avivklas — APPROVED** [review]

**Pull request merged**